src/sys/signal.rs

```diff
@@ -831,6 +831,95 @@ mod sigevent {
     }
 }
 
+/// A wrapper for `stack_t`, used to represent alternate stacks for `sigaltstack`.
+#[derive(Copy, Clone)]
+#[allow(missing_debug_implementations)]
+pub struct SigStack {
+    stack: libc::stack_t,
+}
+
+impl SigStack {
+    /// Define an alternate signal stack for use with `sigaction`.
+    ///
+    /// See [`sigaltstack(3posix)`](http://man7.org/linux/man-pages/man3/sigaltstack.3p.html) for
+    /// details on the requirements for an alternate signal stack.
+    pub fn new(sp: *mut libc::c_void, flags: SigStackFlags, size: libc::size_t) -> SigStack {
+        let mut stack = unsafe { mem::uninitialized::<libc::stack_t>() };
+        stack.ss_sp = sp;
+        stack.ss_flags = flags.bits();
+        stack.ss_size = size;
+        SigStack { stack }
+    }
+
+    pub fn flags(&self) -> SigStackFlags {
+        SigStackFlags::from_bits_truncate(self.stack.ss_flags)
+    }
+}
+
+impl AsRef<libc::stack_t> for SigStack {
+    fn as_ref(&self) -> &libc::stack_t {
+        &self.stack
+    }
+}
+
+impl AsMut<libc::stack_t> for SigStack {
+    fn as_mut(&mut self) -> &mut libc::stack_t {
+        &mut self.stack
+    }
+}
+
+libc_bitflags! {
+    pub struct SigStackFlags: libc::c_int {
+        SS_ONSTACK;
+        SS_DISABLE;
+    }
+}
+
+/// Set an alternate stack for signal handlers on this thread to optionally run on.
+///
+/// See [`sigaltstack(3posix)`](http://man7.org/linux/man-pages/man3/sigaltstack.3p.html) for
+/// details.
+///
+/// # Safety
+///
+/// In addition to the requirements described in the manpage, the current thread must have
+/// read/write access to the memory pointed to by the `SigStack` argument, otherwise any subsequent
+/// signal handlers configured with the `SaFlags::SA_ONSTACK` flag will cause a double fault.
+///
+/// ```
+/// # use nix::libc::SIGSTKSZ;
+/// # use nix::sys::mman::{mmap, MapFlags, ProtFlags};
+/// # use nix::sys::signal::{sigaltstack, SigStack, SigStackFlags};
+/// # use std::ptr;
+/// let stack_mem = unsafe {
+///     mmap(
+///         ptr::null_mut(),
+///         SIGSTKSZ,
+///         ProtFlags::PROT_READ | ProtFlags::PROT_WRITE,
+///         MapFlags::MAP_ANON | MapFlags::MAP_PRIVATE,
+///         -1,
+///         0,
+///     )
+///     .unwrap()
+/// };
+/// let sigstack = SigStack::new(stack_mem, SigStackFlags::empty(), SIGSTKSZ);
```

Comment:
For the docs, `SIGSTKSZ` looks mysterious and undefined.  Better to refer to it as `libc::SIGSTKSZ`.